Add cryptotokenPrivate and device support (hid and usb) for U2F

evq opened the pull request:
This PR contains the following changes required to enable u2f support:
1. Instantiate the chrome device client to enable the existing `chrome.hid` and `chrome.usb` apis
2. Implement `chrome.windows.get` which is used by the cryptotoken extension
3. Build and register the `chrome.cryptotokenPrivate` extension api
4. Allow loading of component extensions from pak resources via the brave extension api

I have tested these changes on linux, mac and windows.

bridiver commented on atom/common/api/resources/windows_bindings.js:
```diff
@@ -17,7 +17,21 @@ binding.registerCustomHook(function (bindingsAPI, extensionId) {
 
   apiFunctions.setHandleRequest('get', function (windowId, getInfo, cb) {
-    console.warn('chrome.windows.get is not supported yet')
+    if (arguments.length === 1) {
```
you actually don't need all of these checks because`setHandleRequest` normalizes the arguments

bridiver commented on atom/common/api/resources/windows_bindings.js:
```diff
@@ -17,7 +17,21 @@ binding.registerCustomHook(function (bindingsAPI, extensionId) {
+      getInfo = arguments[1]
+      cb = arguments[2]
+    }
+    var responseId = ++id
```
we've moved away from using `++id` because it can produce duplicate requests across multiple windows. See `tabs_bindings.js` for usage of `ipc.guid()`

bridiver commented on BUILD.gn:
```diff
@@ -121,6 +122,7 @@ repack("packed_extra_resources") {
     ":atom_resources",
     ":brave_resources",
     "app:brave_strings",
+    "//chrome/browser/resources:component_extension_resources_grit",
```
shouldn't think be `//chrome/browser/resources:component_extension_resources`? I don't see a target for `component_extension_resources_grit`

bridiver commented on atom/browser/extensions/atom_component_extension_resource_manager.cc:
```diff
@@ -0,0 +1,71 @@
+AtomComponentExtensionResourceManager::
+~AtomComponentExtensionResourceManager() {}
+
+bool AtomComponentExtensionResourceManager::IsComponentExtensionResource(
```
most of this code looks identical to `ChromeComponentExtensionResourceManager`. We want to avoid copying chrome code so can you subclass instead?

evq replied:
I don't think I can subclass since `AddComponentResourceEntries` and `path_to_resource_id_` are private

bridiver commented on chromium_src/BUILD.gn:
```diff
@@ -400,10 +404,18 @@ source_set("browser") {
+        "//chrome/browser/extensions/chrome_extension_function_details.h",
         "//chrome/browser/extensions/event_router_forwarder.cc",
         "//chrome/browser/extensions/event_router_forwarder.h",
+        "//chrome/browser/extensions/window_controller.cc",
```
this looks like a potential problem to me because along with windows_util above. None of these methods will work correctly without code so anything that uses them will get the wrong result. I think we need to look at the usage and possibly override with our own implementation

bridiver commented on lib/browser/api/extensions.js:
```diff
@@ -603,6 +603,28 @@ var windowInfo = function (win, populateTabs) {
   }
 }
 
+ipcMain.on('chrome-windows-get', function (evt, responseId, windowId, getInfo) {
```
is this a rebase issue? Doesn't seem related

evq replied:
`chrome.windows.get` is needed by the `cryptotoken` extension, thus I implemented it

alexwykoff commented:
What are the next steps @evq ?

evq commented:
@alexwykoff 
I have some local changes to address the requested changes which I was working on pre-Mercury, I need to rebase and finish those off. Barring major stumbling blocks during rebase, it shouldn't take much time at all once I can get back at it. Would really like to target the end of next week.

evq commented:
@bridiver apologies for the long hiatus on this one... finally got some time to rebase, finish addressing your comments and test!

the primary change avoids copying a bunch of code from `ChromeComponentExtensionResourceManager` - I am now using it directly and have removed the small shim `AtomComponentExtensionResourceManager`. I've created a helper method to enable loading the extension manifest from the resources bundle.

bridiver commented on atom/browser/BUILD.gn:
```diff
@@ -310,6 +310,8 @@ source_set("browser") {
       "extensions/api/atom_extensions_api_client.h",
       "extensions/atom_browser_client_extensions_part.cc",
       "extensions/atom_browser_client_extensions_part.h",
+      "extensions/atom_component_extensions.cc",
```
since this uses brave_resources.h we need to add a dep for `brave_resources` here too

evq commented on atom/browser/BUILD.gn:
```diff
@@ -306,6 +308,8 @@ source_set("browser") {
   }
 
   if (enable_extensions) {
+    data_deps += [ "//electron:brave_resources" ]
```
is this the right way to address your previous comment re: the missing `brave_resources.h` dependency? wasn't sure since it is generated and I'm not particularly familiar with GN

bridiver commented on brave/browser/api/brave_api_extension.cc:
```diff
@@ -98,10 +107,17 @@ scoped_refptr<extensions::Extension> LoadExtension(const base::FilePath& path,
-    return NULL;
+
+  int resource_id;
+  if (!IsComponentExtension(path, &resource_id)) {
```
can we use IsComponentExtensionResource here instead? `IsComponentExtension` is a bit misleading because we have component extensions that are not loaded from resource paks

bridiver commented on muon/browser/muon_browser_process_impl.cc:
```diff
@@ -19,6 +20,8 @@ MuonBrowserProcessImpl::MuonBrowserProcessImpl(
     BrowserProcessImpl(local_state_task_runner, command_line) {
   g_browser_process = this;
+
+  device_client_.reset(new ChromeDeviceClient);
```
this should actually go in chromium_src browser_process_impl.cc
the muon subclass is for muon-specific stuff
